[Tan Joe Wel] iP

tanjoewel opened the pull request:
# Duke
## The task manager you never knew you needed

> "Your mind is for having ideas, not holding them" - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

#### Duke frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* very fast to use

and all you need to do is `while(True)` ~~all this very tedious things~~,

download it from here.
~~Insert download link here~~
1. run the file
1. add your tasks.
1. let it manage your tasks for you 😉

Features:

- [ ] Creating tasks
- [X] Marking tasks as done
- [ ] GUi (Coming soon!)

Now with 
```Java
System.out.println("200% more types of tasks!")
```

Commit:
Add Gradle support

Commit:
Added features for level-1

Commit:
Added the "add" and "list" feature

Commit:
Added functionality to mark tasks as done

Commit:
added the different types of events!

Commit:
Implemented error handling

Commit:
Added delete functionality

Commit:
Add feature to save to hard disk

Rewrite most of the code to make it neater and slightly abiding by OOP principles.

Commit:
Add feature for Duke to recognize date and time

Commit:
Fix some issues

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
Refactor code to make it more OOP

Commit:
Organize into packages

Commit:
Add overriding equals method in various tasks

Add JUnit tests in DukeTest.java

Commit:
Add javadocs for all the methods

Commit:
Tweak code to comply with coding standards

Commit:
Add feature for Duke to find task based on keyword

Commit:
Merge A-JavaDocs, A-CodingStandard and A-Level-9 into master branch

Jeromesyl commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,214 @@
+        try {
+            int i = userInput.indexOf("/");
+            String description = userInput.substring(6,i-1);
+            //String time = userInput.substring(i+1);
```
Should this line be left here? Is it still needed?

Jeromesyl commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    /**
+     * Executes when Duke receives a deadline command
+     * Adds a deadline task to the list of tasks
+     */
+    public void deadline_execute() {
+        try {
+            int i = userInput.indexOf("/");
+            String description = userInput.substring(9,i-1);
+            //String time = userInput.substring(i+1);
+            String[] split = userInput.split("by");
+            String date = split[1].substring(1);
+            LocalDate d = LocalDate.parse(date);
+            String formattedTime = d.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+            Task A = new Deadlines(description, formattedTime);
+            tasks.addTask(A);
+            Duke.index++;
+            String message = "    ----------------------------\n"
+                    + "Got it, I've added this task: \n"
+                    + A.toString() + "\n"
+                    + "Now you have " + Duke.index + " tasks in the list\n"
+                    + "    ----------------------------";
+            System.out.println(message);
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Oops! The description of a deadline cannot be empty and must contain a time!");
+        }
+    }
```
Since Todo, Deadline, and Event all inherit from Task, perhaps it would be better to make this an abstract method in the Task class and implement them in the individual sub classes instead?

Jeromesyl replied:
Or you could just have the method in the Task class and override the method in each subclass.

Jeromesyl commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,64 @@
+        this.isDone = false;
+    }
+
+    public String getTaskIcon() {
+        return "a";
+    }
```
You are missing a javadoc comment here.

Jeromesyl reviewed (commented):
LGTM. Hardly any coding standard violations. I like how organized and easy to read your code is.

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/ToDo.java

Commit:
Add comment to credit Hong Ai Ling for inspiration on how to load tasks from hard disk (sorry for not doing this earlier!)

tanjoewel commented:
Hi Jerome thanks for reviewing my code! Appreciate the input!

Commit:
Modify some of the JavaDoc comments and add a missing JavaDoc comment

Commit:
Install Gradle

Commit:
Automate building and running of project using Gradle

Modify code to comply with CheckStyle (had to do this to successfully build with Gradle)

Commit:
Add functionality to automatically create a file when file "data/Duke.txt" is not found

Commit:
Add basic GUi functionality to Duke

Commit:
Iteration 2: Add dialog box and images

charliemoweng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Scanner;
```
Maybe can separate this from the two above?

charliemoweng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+@SuppressWarnings("checkstyle:Regexp")
+public class Duke {
+
+    private int index = 0; //will figure out a way to make this private :<
```
what is this comment for? isn't it already private?

charliemoweng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+
+        bot.ui.printGreeting();
+
+        while (true) {
```
Maybe can extract this out into a run() method? like suggested in the weekly task page.

charliemoweng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,215 @@
+     * Executes when Duke receives a bye command
+     * Sends a goodbye message to the user
+     */
+    public void bye_execute() {
```
i think for methods need to use camelCase.

charliemoweng reviewed (commented):
Overall good job! Keep up the good work.

Commit:
Small update

Commit:
Small update

Commit:
Add GUi for Duke

Commit:
Change main class to Launcher instead of Duke so that Gradle runs the GUi and not CLI

Commit:
Add assertions to code

Commit:
Refactor code to increase code quality

Code has magic literals for various tasks

Magic literals should be avoided and used in constants instead

Change magic literals to be constants

tanjoewel committed:
Merge pull request #2 from tanjoewel/branch-A-Assertions

Add assertions to code

tanjoewel committed:
Revert "Add assertions to code"

tanjoewel committed:
Merge pull request #4 from tanjoewel/revert-2-branch-A-Assertions

Revert "Add assertions to code"

Commit:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/Duke.java

Commit:
Add functionality for Duke to return statistics.

Duke can add and delete tasks and mark tasks as done, but sometimes the list of tasks get very long and quite unwieldy to handle.

Providing the ability for the user to view statistics allows the user to quickly see their progress at a glance.

To provide such an ability, three new commands have been added with the purpose of showing the user such statistics.

tanjoewel committed:
Update README.md

tanjoewel committed:
Update README.md

Commit:
Change some of the user inputs to all lower case

Commit:
Merge branch 'master' of https://github.com/tanjoewel/ip

tanjoewel committed:
Update README.md

tanjoewel committed:
Add files via upload

Commit:
Rename the screenshot to Ui.png
